Initial code import.

carlosalberto opened the pull request.

carlosalberto committed:
Initial code import.

carlosalberto committed:
Better tracing for pipelines.

* Don't trace call with zero commands
  (these invocations don't even hit the server).
* Trace calls commands that get sent immediately
  from a pipeline, such as 'WATCH'.

carlosalberto committed:
Add more tests for testing pipelines being patched.

carlosalberto committed:
Add simple/explicit examples.

joeblubaugh commented on setup.py:
```diff
@@ -0,0 +1,27 @@
+setup(
+    name='redis_opentracing',
+    version=version,
+    url='https://github.com/carlosalberto/python-redis/',
```
```
url
download_url
description
```

all likely need to change when we do a PR against opentracing-contrib.

carlosalberto replied:
Oh definitely.

joeblubaugh commented on redis_opentracing/__init__.py:
```diff
@@ -0,0 +1,149 @@
+            # Nothing to process/handle.
+            return execute_method(raise_on_error=raise_on_error)
+
+        span = g_tracer.start_span(_get_operation_name('MULTI'))
```
Is there a reasonable method for attaching a function that will give us the parent_span (if any?)

These calls to redis likely have upstream spans associated with a request that hits the server. We ought to figure out a way to make sure they get joined together with the current_span, or some other span context.

carlosalberto replied:
So, the thing about the parent span, I TOTALLY FORGOT to mention as part as the PR: there's a current effort to provide an ActiveSpanSource, which is similar with what happened with Java recently. This is specially important when incorporating tracing with async frameworks (Tornado, gevent, the new asyncio). Once we have something in place, I will add the required hooks this and there.

Observe the other libraries I worked on (elasticsearch, pyramid, sqlalchemy) provide their own parent span calls (mostly on a thread-local basis... which won't be enough - so they will have to be updated eventually, too). Hope that helps :)

joeblubaugh replied:
Sounds good - this is definitely an evolving part of usage, especially when languages support a wide variety of async models.

joeblubaugh commented on examples/explicit.py:
```diff
@@ -0,0 +1,31 @@
+import datetime
+import redis
+
+import lightstep
```
Do we want to directly use LightStep in this tracer example? Can we avoid requiring any OT implementation and just show some kind of "set up your tracer" placeholder?

carlosalberto replied:
So far we have been using the `lightstep` tracer as part of the examples everywhere. One thing we could do is simply use the `opentracing.tracer` instance from here, and put a comment on the examples to set that first. 

Else, at least a comment there, saying `Or replace it here the tracer of your choice`.

joeblubaugh replied:
Yes, I would replace this with a comment.

joeblubaugh commented on redis_opentracing/__init__.py:
```diff
@@ -0,0 +1,149 @@
+g_trace_prefix = None
+g_trace_all_classes = True
+
+def init_tracing(tracer, trace_all_classes=True, prefix='Redis'):
```
Can you add doc strings for all the public defs in this file that adhere to [PEP 287](https://www.python.org/dev/peps/pep-0287/)

In particular, clarifying the role of `trace_all_classes` and `prefix` for users would be great.

carlosalberto replied:
Sure, the plan was to add it today at some point (with a good, clear README as well) :)

joeblubaugh reviewed (commented):
Hi Carlos. I'm Joe, an engineer at LightStep.

At a high level, I think my main question about this code is figuring out a way that we can link these spans with either a parent span from the same process or a remote span whose context has been propagated to this process.

This is a little "up in the air" right now, but both the Django and Flask implementations on opentracing-contrib provide methods to get the span for a given request. Is there a common question interface we can define that will let us access those spans when starting new ones?

carlosalberto committed:
Add a proper README and code documentation.

carlosalberto committed:
Add a setup.cfg file for deployment/publishing.

joeblubaugh commented on README.rst:
```diff
@@ -0,0 +1,65 @@
+
+This package enables distributed tracing for the Python redis library.
+
+Instalation
```
Instalation -> Installation

joeblubaugh commented on README.rst:
```diff
@@ -0,0 +1,65 @@
+Getting started
+===============
+
+Please see the examples directory. Overall, usage requires that a tracer gets set and have the Redis client and pipelines functionality get automatically traced:
```
Maybe rephrase this:

```
Tracing a Redis client requires setting up an OpenTracing-compatible tracer, and calling `init_tracing` to set up the tracing wrappers. See the examples directory for several different approaches.
```

joeblubaugh commented on README.rst:
```diff
@@ -0,0 +1,65 @@
+    client = redis.StrictRedis()
+    client.set('last_access', datetime.datetime.now())
+
+It's possible to have traced only specific `StrictRedis` client objects:
```
Rephrase: `It's possible to trace only specific Redis clients:`

joeblubaugh commented on README.rst:
```diff
@@ -0,0 +1,65 @@
+#################
```
Great README, thanks!

joeblubaugh commented on README.rst:
```diff
@@ -0,0 +1,65 @@
+    pipe.execute()
+
+
+A thing to notice about pipelines is that when executed as a transaction, the set of commands will be included under a single 'MULTI' operation. Else, specially with commands required to be executed immediately, tracing will happen one-by-one (i.e. the `WATCH` command).
```
Rephrase:
```NOTE: when pipeline commands are executed as a transaction, these commands will be grouped under a single 'MULTI' operation. They'll also appear as a single operation in the trace. Outside of a transaction, each command will generate a span.```

joeblubaugh commented on redis_opentracing/__init__.py:
```diff
@@ -0,0 +1,173 @@
+
+def init_tracing(tracer, trace_all_classes=True, prefix='Redis'):
+    '''
+    Set our tracer for Redis. Tracer objects from our
```
replace "our django/flask/pyramid" with "the OpenTracing django/flask/pyramid"

joeblubaugh reviewed (approved):
Very nice overall! Some minor changes to wording in a few of the README and comment sections.

carlosalberto committed:
Update the README with improved wording.

carlosalberto committed:
Don't use lightstep for the examples.

Simply tell the user to use their own
OpenTracing-compatible tracer himself/herself.

carlosalberto committed:
Improve wording for init_tracing() usage.

carlosalberto committed:
Fix typo for Installation (not Instalation).

carlosalberto commented:
Merging after applying your suggestions/feedback, Joe. Greatly appreciated. Will poke you on the next iteration (i.e. once we land the `ActiveSpanSource` and integrate it here).
